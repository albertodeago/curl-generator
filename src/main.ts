import { bodyToCommand, CurlBody } from "./bodies/body";

type StringMap = { [key: string]: string };

/**
 * Additional options for curl command.
 *
 * --compressed        ->   Request compressed response
 * --compressed-ssh    ->   Enable SSH compression
 * --fail              ->   Fail silently (no output at all) on HTTP errors
 * --fail-early        ->   Fail on first transfer error, do not continue
 * --head              ->   Show document info only
 * --include           ->   Include protocol response headers in the output
 * --insecure          ->   Allow insecure server connections when using SSL
 * --ipv4              ->   Resolve names to IPv4 addresses
 * --ipv6              ->   Resolve names to IPv6 addresses
 * --list-only         ->   List only mode
 * --location          ->   Follow redirects
 * --location-trusted  ->   Like --location, and send auth to other hosts
 * --no-keepalive      ->   Disable TCP keepalive on the connection
 * --show-error        ->   Show error even when -s is used
 * --silent            ->   Silent mode
 * --ssl               ->   Try SSL/TLS
 * --sslv2             ->   Use SSLv2
 * --sslv3             ->   Use SSLv3
 * --verbose           ->   Make the operation more talkative
 */
type CurlAdditionalOptions = {
  compressed?:        boolean,
  compressedSsh?:     boolean,
  fail?:              boolean,
  failEarly?:         boolean,
  head?:              boolean,
  include?:           boolean,
  insecure?:          boolean,
  ipv4?:              boolean,
  ipv6?:              boolean,
  listOnly?:          boolean,
  location?:          boolean,
  locationTrusted?:   boolean,
  noKeepalive?:       boolean,
  output?:            string,
  showError?:         boolean,
  silent?:            boolean,
  ssl?:               boolean,
  sslv2?:             boolean,
  sslv3?:             boolean,
  verbose?:           boolean,
};

type CurlRequest = {
  method?: "GET" | "get" | "POST" | "post" | "PUT" | "put" | "PATCH" | "patch" | "DELETE" | "delete",
  headers?: StringMap,
  body?: CurlBody,
  url: string,
};

// slash for connecting previous breakup line to current line for running cURL directly in Command Prompt
const slash = " \\";
const newLine = "\n";

/**
 * @param {string} [method]
 * @returns {string}
 */
const getCurlMethod = function (method?: string): string {
  let result: string = "";
  if (method) {
    const types: StringMap = {
      GET: "-X GET",
      POST: "-X POST",
      PUT: "-X PUT",
      PATCH: "-X PATCH",
      DELETE: "-X DELETE",
    };
    result = ` ${types[method.toUpperCase()]}`;
  }
  return slash + newLine + result;
};

/**
 * @param {StringMap} headers
 * @returns {string}
 */
const getCurlHeaders = function (headers?: StringMap): string {
  let result = "";
  if (headers) {
    Object.keys(headers).map((val) => {
      result += `${slash}${newLine} -H '${val}: ${headers[val].replace(
        /(\\|')/g,
        "\\$1"
      )}'`;
    });
  }
  return result;
};

/**
 * @param {CurlBody} body
 * @returns {string}
 */
const getCurlBody = function (body?: CurlBody): string {
  let result = "";
  if (body) {
    result += `${slash}${newLine} ${bodyToCommand(body)}`;
  }
  return result;
};

/**
 * Given the curl additional options, turn them into curl syntax
 * @param {CurlAdditionalOptions} [options]
 * @returns {string}
 */
const getCurlOptions = function (options?: CurlAdditionalOptions): string {
  let result = "";
  if (options) {
    (Object.keys(options) as Array<keyof CurlAdditionalOptions>).forEach(
      (key: keyof CurlAdditionalOptions) => {
        const kebabKey = key.replace(
          /[A-Z]/g,
          (letter) => `-${letter.toLowerCase()}`
        );

        if (!options[key]) {
          throw new Error(`Invalid Curl option ${key}`);
        } else if (typeof options[key] === "boolean" && options[key]) {
          // boolean option, we just add --opt
          result += ` --${kebabKey}`;
        } else if (typeof options[key] === "string") {
          // string option, we have to add --opt=value
          result += ` --${kebabKey} ${options[key]}`;
        }
      }
    );
  }

  return result ? `${slash}${newLine}${result}` : result;
};

/**
 * @param {CurlRequest} params
 * @param {CurlAdditionalOptions} [options]
 * @returns {string}
 */
const CurlGenerator = function (
  params: CurlRequest,
  options?: CurlAdditionalOptions
): string {
  let curlSnippet = "curl ";
  curlSnippet += params.url;
  curlSnippet += getCurlMethod(params.method);
  curlSnippet += getCurlHeaders(params.headers);
  curlSnippet += getCurlBody(params.body);
  curlSnippet += getCurlOptions(options);
  return curlSnippet.trim();
};

export { CurlGenerator };
