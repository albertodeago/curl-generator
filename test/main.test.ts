import { expect, test } from "vitest";
import { CurlGenerator } from "../src/main";
import * as params from "../example/parameters";

Object.keys(params).forEach((key) => {
  test(key, () => {
    const result = CurlGenerator(params[key], params[key].options);
    expect(result).toMatchSnapshot();
  });
});

test("error body type", () => {
  expect(() =>
    CurlGenerator({
      url: "https://jsonplaceholder.typicode.com/posts",
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: 123 as unknown as string,
    })
  ).toThrowError(/^Invalid body type/);
});

test("error form type", () => {
  expect(() =>
    CurlGenerator({
      url: "https://jsonplaceholder.typicode.com/posts",
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: {
        type: "form",
        content: {
          file1: {
            type: "blob",
          },
        },
      },
    })
  ).toThrowError(/^Invalid form body value type/);
});

test("error options", () => {
  expect(() =>
    CurlGenerator(
      { url: "https://jsonplaceholder.typicode.com/posts/1" },
      {
        silent: false as unknown as true,
      }
    )
  ).toThrowError(/^Invalid Curl option/);
});
