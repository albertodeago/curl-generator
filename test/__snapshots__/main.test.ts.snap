// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`del1 1`] = `
"curl https://jsonplaceholder.typicode.com/posts/1 \\
 -X DELETE"
`;

exports[`fileBody 1`] = `
"curl https://jsonplaceholder.typicode.com/posts \\
 -X POST \\
 -H 'Content-Type: application/json' \\
 --data-binary @data.json"
`;

exports[`formBody1 1`] = `
"curl https://jsonplaceholder.typicode.com/posts \\
 -X POST \\
 -H 'Content-Type: application/x-www-form-urlencoded' \\
 -F name=foo"
`;

exports[`formBody2 1`] = `
"curl https://jsonplaceholder.typicode.com/posts \\
 -X POST \\
 -H 'Content-Type: multipart/form-data' \\
 -F file1=@data.json"
`;

exports[`formBody3 1`] = `
"curl https://jsonplaceholder.typicode.com/posts \\
 -X POST \\
 -H 'Content-Type: application/x-www-form-urlencoded' \\
 -F file1=string data"
`;

exports[`formBody4 1`] = `
"curl https://jsonplaceholder.typicode.com/posts \\
 -X POST \\
 -H 'Content-Type: application/x-www-form-urlencoded' \\
 -d 'key1=value+1'"
`;

exports[`get1 1`] = `"curl https://jsonplaceholder.typicode.com/posts/101 \\"`;

exports[`get2 1`] = `
"curl https://jsonplaceholder.typicode.com/todos/1 \\
 -X GET"
`;

exports[`getEscapeHeaders 1`] = `
"curl https://jsonplaceholder.typicode.com/todos/1 \\
 -X GET \\
 -H 'key: a "strange" value'"
`;

exports[`getWithHeaders 1`] = `
"curl https://jsonplaceholder.typicode.com/todos/1 \\
 -X GET \\
 -H 'Content-Type: application/json'"
`;

exports[`jsonBody 1`] = `
"curl https://jsonplaceholder.typicode.com/posts \\
 -X POST \\
 -H 'Content-Type: application/json' \\
 -d '{"name":"foo"}'"
`;

exports[`opt1 1`] = `
"curl https://jsonplaceholder.typicode.com/posts/1 \\
 \\
 --silent --show-error"
`;

exports[`opt2 1`] = `
"curl https://jsonplaceholder.typicode.com/posts/1 \\
 \\
 --output test.txt --silent"
`;

exports[`patch1 1`] = `
"curl https://jsonplaceholder.typicode.com/posts/1 \\
 -X PATCH \\
 -H 'Content-type: application/json; charset=UTF-8' \\
 -d '{"title":"foo patched"}'"
`;

exports[`post1 1`] = `
"curl https://jsonplaceholder.typicode.com/posts \\
 -X POST \\
 -H 'Content-type: application/json; charset=UTF-8' \\
 -d '{"id":"123-456-789","key1":"value 1","key2":"a \\\\"complex\\\\" value"}'"
`;

exports[`put1 1`] = `
"curl https://jsonplaceholder.typicode.com/posts/1 \\
 -X PUT \\
 -H 'Content-type: application/json; charset=UTF-8' \\
 -d '{"id":1,"title":"foo","body":"barzzz","userId":1}'"
`;

exports[`rawBody1 1`] = `
"curl https://jsonplaceholder.typicode.com/posts \\
 -X POST \\
 -H 'Content-Type: text/plain' \\
 -d 'string data'"
`;

exports[`rawBody2 1`] = `
"curl https://jsonplaceholder.typicode.com/posts \\
 -X POST \\
 -H 'Content-Type: text/plain' \\
 -d 'string data'"
`;

exports[`searchParamsBody 1`] = `
"curl https://jsonplaceholder.typicode.com/posts \\
 -X POST \\
 -H 'Content-Type: text/plain' \\
 -d 'key1=value+1'"
`;
